server/services/maps/custom.js

```diff
@@ -0,0 +1,102 @@
+const ValidationError = require("../../errors/validation");
+
+module.exports = class CustomMapService {
+    constructor() { }
+
+    generateLocations(game, settings, starCount, playerCount) {
+        let json;
+
+        try {
+          json = JSON.parse(settings.galaxy.customJSON)
+        }
+        catch (e) {
+          throw new ValidationError('The custom map JSON is malformed.')
+        }
+
+        const locations = [];
+        const playerIndexes = [];
+        //const nameList = new Set()
+        const homeStars = []
+
+        for (const star of json.stars) {
+            //if (nameList.has(star.name)) continue
+
+            if (!this._checkStarProperty(star?.location, 'x', 'number')) continue
+            if (!this._checkStarProperty(star?.location, 'y', 'number')) continue
+            if (!this._checkStarProperty(star, 'naturalResources', 'number')) continue
+            if (!this._checkStarProperty(star, 'warpGate', 'boolean')) continue
+            if (!this._checkStarProperty(star?.infrastructure, 'economy', 'number')) continue
+            if (!this._checkStarProperty(star?.infrastructure, 'industry', 'number')) continue
+            if (!this._checkStarProperty(star?.infrastructure, 'science', 'number')) continue
+            //if (!this._checkStarProperty(star, 'ships', 'number')) continue
+            if (!this._checkStarProperty(star, 'playerIndex', 'number')) continue
+            if (!this._checkStarProperty(star, 'homeStar', 'boolean')) continue
+            if (!this._checkStarProperty(star, 'specialistId', 'number')) continue
+
+            if (star.naturalResources < 0 || star.naturalResources > 100)
+                throw new ValidationError('Illigal starting amount of resources, range needs to be between 10 and 100 inclusive')
+            if (star.playerIndex >= (8*4)) //colours*shapes
```

Comment:
`8*4` - Can we avoid magic numbers? Count the actual colours and shapes we have from the `colours.json` file?

Reply:
Possible but would the not be better to add a `shape.json` as well?
And write down there the steps as an path2d array?